### PR TITLE
List container updates and testing

### PR DESCRIPTION
**Pull request by dagar**
Bringing in the container updates for https://github.com/PX4/Firmware/pull/11261 separately.

 - support range based for loops
 - add method to remove a node
 - add clear() to empty entire list and delete nodes
 - add empty() helper

### CODE REVIEW

File: src/include/containers/List.hpp [hunk 1]
```diff
@@ -34,11 +34,13 @@
 /**
  * @file List.hpp
  *
- * A linked list.
+ * An intrusive linked list.
  */
 
 #pragma once
 
+#include <stdlib.h>
```
Comment (julianoes) [inline src/include/containers/List.hpp]: Out of interest, what's used here from `stdlib.h`?

File: src/include/containers/List.hpp [hunk 2]
```diff
@@ -64,8 +66,88 @@ class List
...
+
+		while (node != nullptr) {
+			auto next = node->getSibling();
+			delete node;
+			node = next;
+		}
+
+		_head = nullptr;
+	}
+
```
Comment (julianoes) [inline src/include/containers/List.hpp]: Can we get a unit test for this?

Comment (julianoes) [reply src/include/containers/List.hpp]: Nevermind, found it :smile:

File: src/modules/uORB/uORBDeviceMaster.cpp [hunk 3]
```diff
@@ -189,7 +189,7 @@ void uORB::DeviceMaster::addNewDeviceNodes(DeviceNodeStatisticsData **first_node
 		}
 	}
 
-	for (DeviceNode *node = _node_list.getHead(); node != nullptr; node = node->getSibling()) {
+	for (const auto &node : _node_list) {
```
Comment (julianoes) [inline src/modules/uORB/uORBDeviceMaster.cpp]: Awesome, this is so much nicer!

File: src/systemcmds/tests/test_List.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file test_List.cpp
+ * Tests flight path smoothing algorithms.
```
Comment (julianoes) [inline src/systemcmds/tests/test_List.cpp]: Comment wrong.

File: src/systemcmds/tests/test_List.cpp [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+{
+	List<testContainer *> list1;
+
+	// size should be 0 initially
+	ut_compare("size initially 0", list1.size(), 0);
+	ut_assert_true(list1.empty());
+
+	// insert 100
+	for (int i = 0; i < 100; i++) {
+		testContainer *t = new testContainer();
```
Comment (julianoes) [inline src/systemcmds/tests/test_List.cpp]: You do a `new` here. Where is `delete` for it? Does `delete node;` above clean it up correctly?

File: src/include/containers/List.hpp (hunk 1 above)
Comment (dagar) [reply src/include/containers/List.hpp]: size_t

File: src/systemcmds/tests/test_List.cpp (hunk 5 above)
Comment (dagar) [reply src/systemcmds/tests/test_List.cpp]: The `clear()` method (https://github.com/PX4/Firmware/blob/4ee8b972599e33ae54d483e5a6949439932afccf/src/include/containers/List.hpp#L138-L148) also deletes all nodes. See line 100 below.

Due to the way this List is used throughout the system we have to be a bit flexible with ownership, which means explicitly calling clear to empty and delete everything and not doing that in the destructor.

For example ModuleParams (https://github.com/PX4/Firmware/blob/f229444722132a2e71d92d425597faaadcb480c4/src/platforms/px4_module_params.h#L98) has a List of all children to update, but doesn't own them.

File: src/include/containers/List.hpp (hunk 2 above)
Comment (dagar) [reply src/include/containers/List.hpp]: Thanks, next up is `Queue` after this merges.

File: src/systemcmds/tests/test_List.cpp (hunk 4 above)
Comment (dagar) [reply src/systemcmds/tests/test_List.cpp]: Fixed.

**Pull request merged**